<a href="https://colab.research.google.com/github/plee0617/IS-170-Binder/blob/Association-Rules/LeeIS170_MW_HW5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Pang Yia Lee
# HW5

First Data Set - PreferredColor

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
df=pd.read_csv('PreferredColor.csv')
df.head()

Pink, Brown, Blue, Purple
0  Brown, Brown, Blue, Purple
1    Blue, Pink, Purple, Blue
2          Blue, Blue, Purple
3    Pink, Brown, Brown, Pink
4                Pink, Purple

In [ ]:
# Missing a header so we have to create a new one, titled "Chosen Color"
df=pd.read_csv('PreferredColor.csv', names=['Chosen Color'])
df.head() 

Chosen Color
0   Pink, Brown, Blue, Purple
1  Brown, Brown, Blue, Purple
2    Blue, Pink, Purple, Blue
3          Blue, Blue, Purple
4    Pink, Brown, Brown, Pink

In [ ]:
df.shape

(15, 1)

In [ ]:
# Code will convert the dataframe into a list
data=list(df["Chosen Color"].apply(lambda x:x.split(",")))
data

[['Pink', ' Brown', ' Blue', ' Purple'],
 ['Brown', ' Brown', ' Blue', ' Purple'],
 ['Blue', ' Pink', ' Purple', ' Blue'],
 ['Blue', ' Blue', ' Purple'],
 ['Pink', ' Brown', ' Brown', ' Pink'],
 ['Pink', ' Purple'],
 ['Brown', ' Blue', ' Pink', ' Purple'],
 ['Pink', ' Brown', ' Blue', ' Brown'],
 ['Purple', ' Purple', ' Purple'],
 ['Pink', ' Blue', ' Blue', ' Pink'],
 ['Purple', ' Pink', ' Blue', ' Blue'],
 ['Blue', ' Blue', ' Blue', ' Brown'],
 ['Pink', ' Blue', ' Brown', ' Pink'],
 ['Blue', ' Pink', ' Pink', ' Pink'],
 ['Brown', ' Blue', ' Purple', ' Blue']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Any False or True will be read as 0 or 1 instead
df=df.replace(False,0)
df=df.replace(True,1)

df

Blue   Brown   Pink   Purple  Blue  Brown  Pink  Purple
0       1       1      0        1     0      0     1       0
1       1       1      0        1     0      1     0       0
2       1       0      1        1     1      0     0       0
3       1       0      0        1     1      0     0       0
4       0       1      1        0     0      0     1       0
5       0       0      0        1     0      0     1       0
6       1       0      1        1     0      1     0       0
7       1       1      0        0     0      0     1       0
8       0       0      0        1     0      0     0       1
9       1       0      1        0     0      0     1       0
10      1       0      1        0     0      0     0       1
11      1       1      0        0     1      0     0       0
12      1       1      1        0     0      0     1       0
13      0       0      1        0     1      0     0       0
14      1       0      0        1     0      1     0       0

In [ ]:
df=apriori(df,min_support=0.2,use_colnames=True)
df

support                 itemsets
0   0.733333                  ( Blue)
1   0.400000                 ( Brown)
2   0.466667                  ( Pink)
3   0.533333                ( Purple)
4   0.266667                   (Blue)
5   0.200000                  (Brown)
6   0.400000                   (Pink)
7   0.333333          ( Brown,  Blue)
8   0.333333           ( Pink,  Blue)
9   0.400000         ( Blue,  Purple)
10  0.200000            ( Blue, Blue)
11  0.200000           (Brown,  Blue)
12  0.266667            ( Blue, Pink)
13  0.266667           ( Brown, Pink)
14  0.200000            ( Pink, Pink)
15  0.200000         (Brown,  Purple)
16  0.200000    ( Brown,  Blue, Pink)
17  0.200000  (Brown,  Blue,  Purple)

In [ ]:
# Lift measures the correlation between antecedent and consequent when they are bought separately from each other. 
# Lift = 1 means that there is no correlation within the items.
# Lift > 1 means that there is a positive correlation within the items, i.e., products in the items,x and y, are more likely to be bought together.
# Lift < 1 means that there is a negative correlation within the items, i.e., products in items,x and y, are unlikely to be bought together.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents       consequents  antecedent support  \
0           ( Brown)           ( Blue)            0.400000   
1            ( Blue)          ( Brown)            0.733333   
2            ( Blue)         ( Purple)            0.733333   
3          ( Purple)           ( Blue)            0.533333   
4            ( Blue)            (Blue)            0.733333   
5             (Blue)           ( Blue)            0.266667   
6            (Brown)           ( Blue)            0.200000   
7            ( Blue)           (Brown)            0.733333   
8           ( Brown)            (Pink)            0.400000   
9             (Pink)          ( Brown)            0.400000   
10           ( Pink)            (Pink)            0.466667   
11            (Pink)           ( Pink)            0.400000   
12           (Brown)         ( Purple)            0.200000   
13         ( Purple)           (Brown)            0.533333   
14   ( Brown,  Blue)            (Pink)            0.333333   
15    ( Brown, Pink)           ( Blue)            0.266667   
16     ( Blue, Pink)          ( Brown)            0.266667   
17          ( Brown)     ( Blue, Pink)            0.400000   
18           ( Blue)    ( Brown, Pink)            0.733333   
19            (Pink)   ( Brown,  Blue)            0.400000   
20    (Brown,  Blue)         ( Purple)            0.200000   
21  (Brown,  Purple)           ( Blue)            0.200000   
22  ( Blue,  Purple)           (Brown)            0.400000   
23           (Brown)  ( Blue,  Purple)            0.200000   
24           ( Blue)  (Brown,  Purple)            0.733333   
25         ( Purple)    (Brown,  Blue)            0.533333   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.733333  0.333333    0.833333  1.136364  0.040000    1.600000  
1             0.400000  0.333333    0.454545  1.136364  0.040000    1.100000  
2             0.533333  0.400000    0.545455  1.022727  0.008889    1.026667  
3             0.733333  0.400000    0.750000  1.022727  0.008889    1.066667  
4             0.266667  0.200000    0.272727  1.022727  0.004444    1.008333  
5             0.733333  0.200000    0.750000  1.022727  0.004444    1.066667  
6             0.733333  0.200000    1.000000  1.363636  0.053333         inf  
7             0.200000  0.200000    0.272727  1.363636  0.053333    1.100000  
8             0.400000  0.266667    0.666667  1.666667  0.106667    1.800000  
9             0.400000  0.266667    0.666667  1.666667  0.106667    1.800000  
10            0.400000  0.200000    0.428571  1.071429  0.013333    1.050000  
11            0.466667  0.200000    0.500000  1.071429  0.013333    1.066667  
12            0.533333  0.200000    1.000000  1.875000  0.093333         inf  
13            0.200000  0.200000    0.375000  1.875000  0.093333    1.280000  
14            0.400000  0.200000    0.600000  1.500000  0.066667    1.500000  
15            0.733333  0.200000    0.750000  1.022727  0.004444    1.066667  
16            0.400000  0.200000    0.750000  1.875000  0.093333    2.400000  
17            0.266667  0.200000    0.500000  1.875000  0.093333    1.466667  
18            0.266667  0.200000    0.272727  1.022727  0.004444    1.008333  
19            0.333333  0.200000    0.500000  1.500000  0.066667    1.333333  
20            0.533333  0.200000    1.000000  1.875000  0.093333         inf  
21            0.733333  0.200000    1.000000  1.363636  0.053333         inf  
22            0.200000  0.200000    0.500000  2.500000  0.120000    1.600000  
23            0.400000  0.200000    1.000000  2.500000  0.120000         inf  
24            0.200000  0.200000    0.272727  1.363636  0.053333    1.100000  
25            0.200000  0.200000    0.375000  1.875000  0.093333    1.280000

In [ ]:
# The confidence is the probability of seeing the consequent (C) in same transaction with antecedent (A). 
# The confidence value 1 means antecedent and consequent always occur together (100%).
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar

antecedents       consequents  antecedent support  \
0           ( Brown)           ( Blue)            0.400000   
1            ( Pink)           ( Blue)            0.466667   
2          ( Purple)           ( Blue)            0.533333   
3             (Blue)           ( Blue)            0.266667   
4            (Brown)           ( Blue)            0.200000   
5             (Pink)           ( Blue)            0.400000   
6           ( Brown)            (Pink)            0.400000   
7             (Pink)          ( Brown)            0.400000   
8            (Brown)         ( Purple)            0.200000   
9    ( Brown,  Blue)            (Pink)            0.333333   
10    ( Brown, Pink)           ( Blue)            0.266667   
11     ( Blue, Pink)          ( Brown)            0.266667   
12    (Brown,  Blue)         ( Purple)            0.200000   
13  (Brown,  Purple)           ( Blue)            0.200000   
14           (Brown)  ( Blue,  Purple)            0.200000   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.733333  0.333333    0.833333  1.136364  0.040000    1.600000  
1             0.733333  0.333333    0.714286  0.974026 -0.008889    0.933333  
2             0.733333  0.400000    0.750000  1.022727  0.008889    1.066667  
3             0.733333  0.200000    0.750000  1.022727  0.004444    1.066667  
4             0.733333  0.200000    1.000000  1.363636  0.053333         inf  
5             0.733333  0.266667    0.666667  0.909091 -0.026667    0.800000  
6             0.400000  0.266667    0.666667  1.666667  0.106667    1.800000  
7             0.400000  0.266667    0.666667  1.666667  0.106667    1.800000  
8             0.533333  0.200000    1.000000  1.875000  0.093333         inf  
9             0.400000  0.200000    0.600000  1.500000  0.066667    1.500000  
10            0.733333  0.200000    0.750000  1.022727  0.004444    1.066667  
11            0.400000  0.200000    0.750000  1.875000  0.093333    2.400000  
12            0.533333  0.200000    1.000000  1.875000  0.093333         inf  
13            0.733333  0.200000    1.000000  1.363636  0.053333         inf  
14            0.400000  0.200000    1.000000  2.500000  0.120000         inf

From the result shown above, the two items which are the color Brown and Blue has a 0.40 and 0.73 support and a confidence of .83, which means there is an 83% chance that if a person prefers the color Brown, they will also prefer the color Blue.

**Second Data Set - How many classes are some students taking**

In [ ]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
df=pd.read_csv('Classes.csv')
df.head()

4, 5, 5, 5
0  5, 4, 5, 6
1  1, 4, 5, 6
2     5, 4, 5
3  2, 4, 6, 5
4  4, 4, 2, 4

In [ ]:
# Missing a header so we have to create a new one, titled "Chosen Color"
df=pd.read_csv('Classes.csv', names=['Amount of classes'])
df.head() 

Amount of classes
0        4, 5, 5, 5
1        5, 4, 5, 6
2        1, 4, 5, 6
3           5, 4, 5
4        2, 4, 6, 5

In [ ]:
# check data size/list
df.shape

(15, 1)

In [ ]:
# Code will convert the dataframe into a list
data=list(df["Amount of classes"].apply(lambda x:x.split(",")))
data

[['4', ' 5', ' 5', ' 5'],
 ['5', ' 4', ' 5', ' 6'],
 ['1', ' 4', ' 5', ' 6'],
 ['5', ' 4', ' 5'],
 ['2', ' 4', ' 6', ' 5'],
 ['4', ' 4', ' 2', ' 4'],
 ['5', ' 5', ' 4', ' 6'],
 ['5', ' 5', ' 5', ' 4'],
 ['1', ' 6', ' 4', ' 4'],
 ['5', ' 4', ' 5', ' 3'],
 ['3', ' 5', ' 5', ' 6'],
 ['4', ' 3', ' 4', ' 5'],
 ['5', ' 4', ' 4', ' 6'],
 ['2', ' 4', ' 5', ' 5'],
 ['3', ' 5', ' 4', ' 5']]

In [ ]:
from mlxtend.preprocessing import TransactionEncoder
a=TransactionEncoder()
a_data=a.fit(data).transform(data)
df=pd.DataFrame(a_data, columns=a.columns_)

#Any False or True will be read as 0 or 1 instead
df=df.replace(False,0)
df=df.replace(True,1)

df

2   3   4   5   6  1  2  3  4  5
0    0   0   0   1   0  0  0  0  1  0
1    0   0   1   1   1  0  0  0  0  1
2    0   0   1   1   1  1  0  0  0  0
3    0   0   1   1   0  0  0  0  0  1
4    0   0   1   1   1  0  1  0  0  0
5    1   0   1   0   0  0  0  0  1  0
6    0   0   1   1   1  0  0  0  0  1
7    0   0   1   1   0  0  0  0  0  1
8    0   0   1   0   1  1  0  0  0  0
9    0   1   1   1   0  0  0  0  0  1
10   0   0   0   1   1  0  0  1  0  0
11   0   1   1   1   0  0  0  0  1  0
12   0   0   1   0   1  0  0  0  0  1
13   0   0   1   1   0  0  1  0  0  0
14   0   0   1   1   0  0  0  1  0  0

In [ ]:
df=apriori(df,min_support=0.2,use_colnames=True)
df

support      itemsets
0   0.866667          ( 4)
1   0.800000          ( 5)
2   0.466667          ( 6)
3   0.200000           (4)
4   0.400000           (5)
5   0.666667      ( 5,  4)
6   0.400000      ( 6,  4)
7   0.400000       (5,  4)
8   0.333333      ( 5,  6)
9   0.333333       (5,  5)
10  0.200000       (5,  6)
11  0.266667  ( 5,  6,  4)
12  0.333333   (5,  5,  4)
13  0.200000   (5,  6,  4)

In [ ]:
# Lift measures the correlation between antecedent and consequent when they are bought separately from each other. 
# Lift = 1 means that there is no correlation within the items.
# Lift > 1 means that there is a positive correlation within the items, i.e., products in the items,x and y, are more likely to be bought together.
# Lift < 1 means that there is a negative correlation within the items, i.e., products in items,x and y, are unlikely to be bought together.
df_ar=association_rules(df,metric="lift", min_threshold=1)
df_ar

antecedents consequents  antecedent support  consequent support   support  \
0          (5)        ( 4)            0.400000            0.866667  0.400000   
1         ( 4)         (5)            0.866667            0.400000  0.400000   
2          (5)        ( 5)            0.400000            0.800000  0.333333   
3         ( 5)         (5)            0.800000            0.400000  0.333333   
4          (5)        ( 6)            0.400000            0.466667  0.200000   
5         ( 6)         (5)            0.466667            0.400000  0.200000   
6      (5,  5)        ( 4)            0.333333            0.866667  0.333333   
7      (5,  4)        ( 5)            0.400000            0.800000  0.333333   
8     ( 5,  4)         (5)            0.666667            0.400000  0.333333   
9          (5)    ( 5,  4)            0.400000            0.666667  0.333333   
10        ( 5)     (5,  4)            0.800000            0.400000  0.333333   
11        ( 4)     (5,  5)            0.866667            0.333333  0.333333   
12     (5,  6)        ( 4)            0.200000            0.866667  0.200000   
13     (5,  4)        ( 6)            0.400000            0.466667  0.200000   
14    ( 6,  4)         (5)            0.400000            0.400000  0.200000   
15         (5)    ( 6,  4)            0.400000            0.400000  0.200000   
16        ( 6)     (5,  4)            0.466667            0.400000  0.200000   
17        ( 4)     (5,  6)            0.866667            0.200000  0.200000   

    confidence      lift  leverage  conviction  
0     1.000000  1.153846  0.053333         inf  
1     0.461538  1.153846  0.053333    1.114286  
2     0.833333  1.041667  0.013333    1.200000  
3     0.416667  1.041667  0.013333    1.028571  
4     0.500000  1.071429  0.013333    1.066667  
5     0.428571  1.071429  0.013333    1.050000  
6     1.000000  1.153846  0.044444         inf  
7     0.833333  1.041667  0.013333    1.200000  
8     0.500000  1.250000  0.066667    1.200000  
9     0.833333  1.250000  0.066667    2.000000  
10    0.416667  1.041667  0.013333    1.028571  
11    0.384615  1.153846  0.044444    1.083333  
12    1.000000  1.153846  0.026667         inf  
13    0.500000  1.071429  0.013333    1.066667  
14    0.500000  1.250000  0.040000    1.200000  
15    0.500000  1.250000  0.040000    1.200000  
16    0.428571  1.071429  0.013333    1.050000  
17    0.230769  1.153846  0.026667    1.040000

In [ ]:
# The confidence is the probability of seeing the consequent (C) in same transaction with antecedent (A). 
# The confidence value 1 means antecedent and consequent always occur together (100%).
df_ar=association_rules(df,metric="confidence", min_threshold=0.6)
df_ar

antecedents consequents  antecedent support  consequent support   support  \
0         ( 5)        ( 4)            0.800000            0.866667  0.666667   
1         ( 4)        ( 5)            0.866667            0.800000  0.666667   
2         ( 6)        ( 4)            0.466667            0.866667  0.400000   
3          (5)        ( 4)            0.400000            0.866667  0.400000   
4         ( 6)        ( 5)            0.466667            0.800000  0.333333   
5          (5)        ( 5)            0.400000            0.800000  0.333333   
6     ( 5,  6)        ( 4)            0.333333            0.866667  0.266667   
7     ( 6,  4)        ( 5)            0.400000            0.800000  0.266667   
8      (5,  5)        ( 4)            0.333333            0.866667  0.333333   
9      (5,  4)        ( 5)            0.400000            0.800000  0.333333   
10         (5)    ( 5,  4)            0.400000            0.666667  0.333333   
11     (5,  6)        ( 4)            0.200000            0.866667  0.200000   

    confidence      lift  leverage  conviction  
0     0.833333  0.961538 -0.026667    0.800000  
1     0.769231  0.961538 -0.026667    0.866667  
2     0.857143  0.989011 -0.004444    0.933333  
3     1.000000  1.153846  0.053333         inf  
4     0.714286  0.892857 -0.040000    0.700000  
5     0.833333  1.041667  0.013333    1.200000  
6     0.800000  0.923077 -0.022222    0.666667  
7     0.666667  0.833333 -0.053333    0.600000  
8     1.000000  1.153846  0.044444         inf  
9     0.833333  1.041667  0.013333    1.200000  
10    0.833333  1.250000  0.066667    2.000000  
11    1.000000  1.153846  0.026667         inf

From the result shown above, the two items which are the number 5 and 5 has a 0.80 and 0.86 support and a confidence of .83, which means there is an 83% chance that if a person has taken 5 classes, they will also prefer to take 4 classes as well; or that 5 and 4 classes are the most common.